bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdHandler.java

```diff
@@ -0,0 +1,622 @@
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.imageio.ImageIO;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.eclipse.smarthome.core.i18n.TimeZoneProvider;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.eclipse.smarthome.io.net.http.HttpRequestBuilder;
+import org.openhab.binding.doorbird.internal.listener.DoorbirdUdpListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DoorbirdHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@NonNullByDefault
+public class DoorbirdHandler extends BaseThingHandler {
+    private static final long API_REQUEST_TIMEOUT = 3L;
+    private static final long MONTAGE_UPDATE_DELAY = 5L;
+
+    // Maximum number of doorbell and motion history images stored on Doorbird backend
+    private static final int MAX_HISTORY_IMAGES = 50;
+
+    private final Logger logger = LoggerFactory.getLogger(DoorbirdHandler.class);
+
+    // Get a dedicated threadpool for the long-running listener thread
+    private final ScheduledExecutorService doorbirdScheduler = ThreadPoolManager.getScheduledPool("doorbirdHandler");
+    private @Nullable ScheduledFuture<?> listenerJob;
+    private DoorbirdUdpListener udpListener;
+
+    private @Nullable ScheduledFuture<?> imageRefreshJob;
+    private @Nullable ScheduledFuture<?> doorbellOffJob;
+    private @Nullable ScheduledFuture<?> motionOffJob;
+
+    private @NonNullByDefault({}) DoorbirdConfiguration config;
```

Comment:
This has to be annotated with `@Nullable` too.

```suggestion
    private @Nullable DoorbirdConfiguration config;
```

Reply:
My understanding of the last discussion (which unfortunately I can‘t find ATM) was that „I know this to be NonNull before use by design“ (which is IMO the case here) also qualifies for `@NonNullByDefault({})`.

Reply:
Let me quote something from our sorely missed OHC maintainer (see https://github.com/openhab/openhab-core/pull/910#discussion_r300031597):

> We introduced the usage of `@NonNullByDefault({})` for variables we know that they are not null but this cannot be detected by the tooling. One common example has been references of components that are injected by the declarative service implementation on activation.
For this use case there is no need anymore because every mandatory reference can now be set by using constructor injection.
For optional references, we should already use `@Nullable` because they can be valid null.
Another use case are members that are injected by reflection because of specific deserialization (e.g. Gson).
>
> Why do we need it in this case.
> Wouldn't it be correct to mark these variable `@Nullable`?

Reply:
I saw other bindings do this, so I thought it was ok. For example, 😉 
https://github.com/openhab/openhab2-addons/blob/9d5b40bff36242cd17757e219b4f660831b9ca8a/bundles/org.openhab.binding.darksky/src/main/java/org/openhab/binding/darksky/internal/handler/DarkSkyAPIHandler.java#L64

I'll do whatever you guys want. But, won't marking it nullable require null checks wherever its referenced?

Reply:
Touché! Fair enough. I used it in the DarkSky binding and elsewhere for the same reason like you. Because it worked and no one complained about it. But rules and common usage may change over time ...

> But, won't marking it nullable require null checks wherever its referenced?

Yes, it does. And that is the way to go because it could be null and we are working in an multi threaded environment where such things could result in NPEs.

Another option could be to set the `config` property in line 90 a non null value:

```suggestion
    private DoorbirdConfiguration config = new DoorbirdConfiguration();
```

Reply:
I still don‘t understand how this should happen. The framework should never call handler-methods on an uninitialized thing. And initialize() guarantees that this is NoNull when it returns. There is no code that re-sets this to null.

Reply:
Right, `config` is set in one and only one place (i.e. in `initialize`). And, `getConfigAs` is not defined as `@Nullable` in `BaseThingHandler`, so it's guaranteed to return non-null. I don't see where `config` can ever be null once the thing is initialized.

Reply:
Alright. I did not want to go beyond the scope and screw up everything. I agree, that we can be sure that `config` cannot be null. It is up to you if you want to change it or leave it like it is.